### ISSUE TITLE
Fully support <additional_rule_separator>.

### ISSUE THREAD

**Issue by ypid** [ypid/opening_hours.js#86]
Related to: #53, #75.

The problem is known for some time now, unfortunattly there was no corresponding issue yet. It has also been reported against JSOM, see issue [11538](https://josm.openstreetmap.de/ticket/11538).

Test case: [`Mo-Fri: 7:00-18:00, Sat: 7:00-13:00`](http://openingh.openstreetmap.de/evaluation_tool/?EXP=Mo-Fri%3A%207%3A00-18%3A00%2C%20Sat%3A%207%3A00-13%3A00&DATE=1433754180000&lat=48.7769&lon=9.1844&mode=0)

**Comment by d1g** [comment]
FYI I implemented additional_rule_separator but not as rule_modifier (one of top level rules)
but [as special case in weekday_selector](https://github.com/opening-hours/opening_hours_grammar/commit/29deda5b761a57f8ee5b27973b83cb90dc9f8e22#diff-32eff7202231d3d730735d47a3e6e6a1R131) (lines 131-141)

* , http://openingh.openstreetmap.de/evaluation_tool/?EXP=Mo-Fr%2007%3A00-18%3A00%2C%20Sa%2007%3A00-13%3A00
* ; http://openingh.openstreetmap.de/evaluation_tool/?EXP=Mo-Fr%2007%3A00-18%3A00%3B%20Sa%2007%3A00-13%3A00

Also, to me, `12:00-14:00, We 16:00-18:00` is just non-canonical version of `We 16:00-18:00, 12:00-14:00` (since grammar recognizes latter version, I'm not going to include special case for every of permutations)

**Comment by ypid** [comment]
Right on :wink: I guess you know that the additional_rule_separator is the reason why I considered using a proper formal syntax specification and parser generators in the first place in the hope that with this support it would be easier get this fully covered.

`12:00-14:00, We 16:00-18:00` has many valid uses and I find it acceptable. `We 16:00-18:00, 12:00-14:00` has a different meaning.

Regarding the `Mo-Fr 07:00-18:00; Sa 07:00-13:00` vs. `Mo-Fr 07:00-18:00, Sa 07:00-13:00`. They are equal as the [oh compare feature](#116) can tell you. `Mo-Fr 07:00-18:00; Sa 07:00-13:00` is preferred and using `Mo-Fr 07:00-18:00, Sa 07:00-13:00` simply means someone did not understand the syntax.

**Comment by d1g** [comment]
@ypid, yes but [current interpretation](http://openingh.openstreetmap.de/evaluation_tool/?EXP=12%3A00-14%3A00) of `12:00-14:00` as `12:00-14:00` = `12:00-14:00 Su-Sa` is overly optimistic because usually OH values like HH:MM mean that _this time_ was specified for weekdays (12345, not 67); `12:00-14:00` sometimes will lead you to the closed door during weekends (67), because weekdays were imprecise in OH tag.

Not that this is "wrong" or "right"; but this behaviour should be optional or at least configurable. Maybe there regions where this is irrelevant. But in some regions there clear difference between 12345 and 67 schedules.

**Comment by ypid** [comment]
Now I see where you are going with this. You are absolutely right. `12:00-14:00, We 16:00-18:00` is syntacticly valid but as you said, is not explicit. We can not make this invalid however as it follows the specs. A more reasonable value would be something like: `Mo-Fr 12:00-14:00, We 16:00-18:00; PH off`.

> But in some regions there clear difference between 12345 and 67 schedules.

Sure there are differences. A rule with only a time selector will always match every weekday, regardless of any other circumstance. I had a similar discussions about this during the [default PH off](https://wiki.openstreetmap.org/wiki/Proposed_features/opening_hours_default_PH_off) proposal and I think that trying to be overly smart in an already overly complex syntax (as most mappers say) would be suboptimal.

**Comment by d1g** [comment]
>  We can not make this invalid however as it follows the specs.

You simply cannot interpret ambiguous input. You can give user a warning "incomplete data", but you cannot use this date in any reasonable manner. On some inputs, you cannot even tokenize it.

`PROGRAMMKEYWORD: programm` 
`STRING: [ a-z]+` 

`input : STRING | PROGRAMMKEYWORD`

For input sting `prog` how do you tokenize input? Why?
For input sting `programm` how do you tokenize input? Why?
For input sting `programm programm` how do you tokenize input? Why?

There a reason why interpreters and compilers throw an error and exit immediately instead of trying to correct a program (or OH string) how user would like to think is "correct".

> overly smart in an already overly complex syntax

I have feeling that most users tend to say that all necessary data should be explicit. Personally I prefer any "rule" as explicit data in OSM.

Reason is that `PH off` or `PH on` is not in widespread use because tooling wasn't perfect in this aspect or simply not informed as well as they should (long write-only wiki pages nobody really reads afterwards).

**Comment by ypid** [comment]
> You simply cannot interpret ambiguous input. You can give user a warning "incomplete data", but you cannot use this date in any reasonable manner. On some inputs, you cannot even tokenize it.

I guess you will need to have the discussion then of changing the \<additional_rule_separator\> token in the oh syntax.

In your example, shouldn't the `PROGRAMMKEYWORD` token match first because it was first mentioned in the grammar?

```
PROGRAMMKEYWORD: programm
STRING: [ a-z]+
```

> Personally I prefer any "rule" as explicit data in OSM.

I do too.

> simply not informed as well as they should

At least in the German OSM community, I hope I have addressed this a bit by the [weekly task about opening_hours](https://blog.openstreetmap.de/blog/2015/04/wochenaufgabe-oeffnungszeiten/) last year.

**Comment by d1g** [comment]
> In your example, shouldn't the PROGRAMMKEYWORD token match first because it was first mentioned in the grammar?

Yes, this is how ANTLR behaves by default, but no matter what grammar you build, you cannot execute `prog` input (4 characters) ([read my updated reply](https://github.com/opening-hours/opening_hours.js/issues/86#issuecomment-214881073) about "minimal executable program").

We don't want to simply parse OH rules, we want to **compute** them and preferably by computers.and without guessing.

> At least in the German OSM community, I hope I have addressed this a bit by the weekly task about opening_hours last year.

Yes, but language barriers kicks in, as example [Russian internet is silent about this](https://www.google.ru/search?q=https://blog.openstreetmap.de/blog/2015/04/wochenaufgabe-oeffnungszeiten/ site:.ru) or least disrespects sources.

**Comment by ypid** [comment]
> Yes, but language barriers kicks in, as example Russian internet is silent about this or least disrespects sources.

One reason why I prefer to write everything in English when possible.

> So solution is "Consider adding missing PH specification please" to end user(s).

What do you mean by that? [Key:opening_hours](https://wiki.openstreetmap.org/wiki/Key:opening_hours) doesn’t look too bad.

> But OH is worse than any programming language, because SPEC is written after all programs in the world (OH values) were written.

In a perfect world, yes. But OSM is more like evolution. It [develops](https://wiki.openstreetmap.org/wiki/Talk:Key:opening_hours). I am not sure one could have done a prefect job when designing the spec beforehand but I agree that some inconsistencies could have been avoided that way. But I hope that we can still push in that direction.

> There a reason why (sometimes) interpreters and compilers throw an error and exit immediately instead of trying to correct a program (or OH string) how user would like to think is "correct".

I think there is a need to differentiate between the use case for the oh syntax and this lib in OSM and the way one would ideally solve such a problem in computer science. As of the way the syntax developed, I think you had to at least try to evaluate oh values even if they not follow the spec 100 %. I allowed this with this lib and tried to push towards 100 % spec compatible values by warnings and prettified values.

You are very welcome to work on these issues or even create a 2.0 redesigned spec. I would appreciate that.

**Comment by d1g** [comment]
> >   So solution is "Consider adding missing PH specification please" to end user(s).

> What do you mean by that? Key:opening_hours doesn’t look too bad.
Yes but this isn't shown in editors by default, so only most disciplined OSM contributors who read wikis will (probably) follow this notice.

I meant that support should be in editors (not just as agreement at wiki).

>  I think you had to at least try to evaluate oh values even if they not follow the spec 100 %.

For best user experience, user(s) should be warned about "approximations" during process. And as I said, tumbler _to disable guessing_ will save the day of somebody.

On type mismatch or syntax error most languages and run times simply stop execution, but OH domain is less critical, worst thing we could do is to waste users hours guiding them to closed POIs instead of opened.

Not that I advocate that we should do as was said in textbook, but from practical perspective, we need 2 use cases:
* strict evaluation that never fails (in routing and everywhere)
* approximate evaluation to enter more data or correct values

**Comment by ypid** [comment]
> I meant that support should be in editors (not just as agreement at wiki).

I agree. Go for in when you have a particular editor in mind. Related: https://github.com/openstreetmap/iD/issues/974 and https://josm.openstreetmap.de/ticket/9157

> And as I said, button to disable guessing will save the day of somebody.

You are right. Thats actually a [feature request for the lib](https://github.com/opening-hours/opening_hours.js/issues/160).

**Comment by SlowMo24** [comment]
Hi,

first things first: great work!

Question: Is there any possibility to add a kind of "parse-error" for all values that could not be parsed asking the user to make her value parseable in the first place. 
E.g. `opening_hours=Mo-Fr 10:00-11:00, PH` is, of course, wrong but while the website https://openingh.openstreetmap.de/evaluation_tool/ returns no result at all JOSM throws a "ScriptException: TypeError:"  `Cannot read property "0" from undefined` internally, not reporting any error to the general user.

i'm sorry if this has been mentioned already, I was not able to fully follow all discussions in all related issues.

**Comment by ypid** [comment]
Hi

Thanks that you like it :) So the short version. It is definitely possible, the only issue is that nobody looked into this for quite some time. I will do this eventually but now right now.
